src/rustup/settings.rs

```diff
@@ -0,0 +1,146 @@
+    }
+    fn read_settings(&self) -> Result<Settings> {
+        if !utils::is_file(&self.0) {
+            try!(self.write_settings(Default::default()));
+        }
+        let content = try!(utils::read_file("settings", &self.0));
+        Settings::parse(&content)
+    }
+    pub fn with<T, F: FnOnce(&Settings) -> Result<T>>(&self, f: F) -> Result<T> {
+        f(&try!(self.read_settings()))
+    }
+    pub fn with_mut<T, F: FnOnce(&mut Settings) -> Result<T>>(&self, f: F) -> Result<T> {
+        let mut settings = try!(self.read_settings());
+        let result = try!(f(&mut settings));
+        try!(self.write_settings(settings));
+        Ok(result)
+    }
+}
+
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Settings {
+    pub version: String,
+    pub default_toolchain: Option<String>,
+    pub overrides: BTreeMap<String, String>,
+}
+
+impl Default for Settings {
+    fn default() -> Self {
+        Settings {
+            version: DEFAULT_METADATA_VERSION.to_owned(),
+            default_toolchain: None,
+            overrides: BTreeMap::new()
+        }
+    }
+}
+
+impl Settings {
+    fn path_to_key(path: &Path, notify_handler: NotifyHandler) -> String {
+        utils::canonicalize_path(path, ntfy!(&notify_handler))
+               .display()
+               .to_string()
+    }
+
+    pub fn remove_override(&mut self, path: &Path, notify_handler: NotifyHandler) -> bool {
+        let key = Self::path_to_key(path, notify_handler);
+        self.overrides.remove(&key).is_some()
+    }
+
+    pub fn add_override(&mut self, path: &Path, toolchain: String, notify_handler: NotifyHandler) {
+        let key = Self::path_to_key(path, notify_handler);
+        notify_handler.call(Notification::SetOverrideToolchain(path, &toolchain));
+        self.overrides.insert(key, toolchain);
+    }
+
+    pub fn find_override(&self, dir_unresolved: &Path, notify_handler: NotifyHandler)
+            -> Option<(String, PathBuf)> {
+        let dir = utils::canonicalize_path(dir_unresolved, ntfy!(&notify_handler));
+        let mut maybe_path = Some(&*dir);
+        while let Some(path) = maybe_path {
+            let key = Self::path_to_key(path, notify_handler);
+            if let Some(toolchain) = self.overrides.get(&key) {
+                return Some((toolchain.to_owned(), path.to_owned()));
+            }
+            maybe_path = path.parent();
+        }
+        None
+    }
+
+    pub fn parse(data: &str) -> Result<Self> {
+        let mut parser = toml::Parser::new(data);
+        let value = try!(parser.parse().ok_or_else(move || ErrorKind::ParsingSettings(parser.errors)));
+
+        Self::from_toml(value, "")
+    }
+    pub fn stringify(self) -> String {
+        toml::Value::Table(self.to_toml()).to_string()
+    }
+
+    pub fn from_toml(mut table: toml::Table, path: &str) -> Result<Self> {
+        let version = try!(get_string(&mut table, "version", path));
+        if !SUPPORTED_METADATA_VERSIONS.contains(&&*version) {

```

Comment:
The upgrade to `settings.toml` is orthogonal to the metadata version: since it's lossless it gets converted automatically, so if you upgrade from an old version of multirust to rustup, it will result in a `settings.toml` file with a version of 2, and you will then be required to do a metadata upgrade.